From f621686d8bff1cf2213a3a5876c51a5f28f1f4bf Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Tue, 12 Dec 2023 18:20:01 +0100
Subject: [PATCH 448/841] fix(stm32mp2): update pwr_domain_suspend

Use the new ops pwr_domain_validate_suspend introduced by commit
d34886140c74 ("fix(psci): add optional pwr_domain_validate_suspend
to plat_psci_ops_t")

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 953e6fd5d19e8c8b73b7fba177b16ce798d65e11 fix(stm32mp2): update pwr_domain_suspend]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Ibcadad8deddd1122e2f8396623d5fb6e36bd867b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/347272
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/stm32mp2_pm.c | 38 +++++++++++++++++++++++-----------
 1 file changed, 26 insertions(+), 12 deletions(-)

diff --git a/plat/st/stm32mp2/stm32mp2_pm.c b/plat/st/stm32mp2/stm32mp2_pm.c
index f1700823a..73fa4c9ec 100644
--- a/plat/st/stm32mp2/stm32mp2_pm.c
+++ b/plat/st/stm32mp2/stm32mp2_pm.c
@@ -323,6 +323,31 @@ static bool stm32_freeze_other_core(unsigned int core_id)
 	return result;
 }
 
+static int stm32_pwr_domain_validate_suspend(const psci_power_state_t *target_state)
+{
+	uintptr_t pwr_base = stm32mp_pwr_base();
+	uint32_t stateid = stm32_get_stateid(target_state->pwr_domain_state);
+	u_register_t mpidr = read_mpidr();
+	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
+
+	/* If retention only at D1 level return as nothing is to be done */
+	if (stateid == PWRSTATE_RUN) {
+		return PSCI_E_SUCCESS;
+	}
+
+	/* If CPU2 is not in reset: low power mode is not supported by CPU1 */
+	if ((mmio_read_32(pwr_base + PWR_CPU2D2SR) & PWR_CPU2D2SR_CSTATE_MASK) != 0U) {
+		WARN("PSCI power domain supend request with Cortex M33 running.\n");
+		return PSCI_E_INVALID_PARAMS;
+	}
+
+	if (!stm32_freeze_other_core(core_id)) {
+		return PSCI_E_DENIED;
+	}
+
+	return PSCI_E_SUCCESS;
+}
+
 /* Display Stop2 or Standby1 modes, when console is available */
 static void print_mode_info(const char *mode)
 {
@@ -341,24 +366,12 @@ static void stm32_pwr_domain_suspend(const psci_power_state_t *target_state)
 	uintptr_t rcc_base = stm32mp_rcc_base();
 	bool standby = false;
 	uint32_t stateid = stm32_get_stateid(target_state->pwr_domain_state);
-	u_register_t mpidr = read_mpidr();
-	unsigned int core_id = MPIDR_AFFLVL0_VAL(mpidr);
 
 	/* If retention only at D1 level return as nothing is to be done */
 	if (stateid == PWRSTATE_RUN) {
 		return;
 	}
 
-	/* If CPU2 is not in reset: low power mode is not supported by CPU1 */
-	if ((mmio_read_32(pwr_base + PWR_CPU2D2SR) & PWR_CPU2D2SR_CSTATE_MASK) != 0U) {
-		WARN("PSCI power domain supend request with Cortex M33 running.\n");
-		return;
-	}
-
-	if (!stm32_freeze_other_core(core_id)) {
-		return;
-	}
-
 	/* Request STOP for both cores */
 	mmio_write_32(rcc_base + RCC_C1SREQSETR, RCC_C1SREQSETR_STPREQ_MASK);
 
@@ -806,6 +819,7 @@ static const plat_psci_ops_t stm32_psci_ops = {
 	.cpu_standby = stm32_cpu_standby,
 	.pwr_domain_on = stm32_pwr_domain_on,
 	.pwr_domain_off = stm32_pwr_domain_off,
+	.pwr_domain_validate_suspend = stm32_pwr_domain_validate_suspend,
 	.pwr_domain_suspend = stm32_pwr_domain_suspend,
 	.pwr_domain_on_finish = stm32_pwr_domain_on_finish,
 	.pwr_domain_suspend_finish = stm32_pwr_domain_suspend_finish,
-- 
2.47.2

